src/Tests/RedirectUILanguageTest.php

```diff
@@ -22,4 +23,47 @@ class RedirectUILanguageTest extends RedirectUITest {
    */
   public static $modules = ['redirect', 'node', 'path', 'dblog', 'views', 'taxonomy', 'language', 'content_translation'];
 
+  /**
+   * Tests the fix 404 pages workflow.
+   */
+  public function testFix404RedirectList() {
+    $this->drupalLogin($this->drupalCreateUser(array(
+      'administer redirects',
+      'administer languages',
+    )));
+
+    // Add predefined language.
+    $this->drupalPostForm('admin/config/regional/language/add', array('predefined_langcode' => 'fr'), t('Add language'));
+    $this->assertText('French');
+
+    // Visit a non existing page to have the 404 redirect_error entry.
+    $this->drupalGet('fr/testing');
+
+    $redirect = db_select('redirect_error')->fields('redirect_error')->condition('source', 'fr/testing')->execute()->fetchAll();
+    if (count($redirect) == 0) {
+      $this->fail('No record was added');
+    }
+
+    // Go to the "fix 404" page and check the listing.
+    $this->drupalGet('admin/config/search/redirect/404');
+    $this->assertText('fr/testing');
+    $this->assertText('French');
+    $this->clickLink(t('Add redirect'));
+
+    // Check if we generate correct Add redirect url and if the form is
+    // pre-filled.
+    $destination = Url::fromUri('base:admin/config/search/redirect/404')->toString();
+    $this->assertUrl('admin/config/search/redirect/add', ['query' => ['source' => 'fr/testing', 'language' => 'fr', 'destination' => $destination]]);
+    $this->assertFieldByName('redirect_source[0][path]', 'fr/testing');
+    $this->assertOptionSelected('edit-language-0-value', 'fr');
+
+    // Save the redirect.
+    $this->drupalPostForm(NULL, array('redirect_redirect[0][uri]' => '/node'), t('Save'));
+    $this->assertUrl('admin/config/search/redirect/404');
+
+    // Check if the redirect works as expected.
+    $this->drupalGet('admin/config/search/redirect');
+    $this->drupalGet('fr/testing');
+    $this->assertUrl('fr/node');
```

Comment:
Interesting, so a target /node with the language fr selected ends up on fr/node.
That leads to a possible limitation. A node that have previously been saved with the wrong language ending up with an alias (people link to it) and then we fix the language... The alias changes and there should be a redirect from the old language version to the new one. I fear this either leads to broken redirect target or no redirect at all.
I think the target should be language prefixed in the database and the language selection should only apply to the source matching. A target /node will end up on /node even if the language is set to (apply only) on fr. Otherwise we need to add (optional) support if the target is in the same language context or possibly a different one (like a second language selector?)